### ISSUE TITLE
Ability to accelerate the building of my package?

### ISSUE THREAD

**Issue by vemv** [melpa/melpa#8741]
Hi!

Sometimes I fix things for a specific user that may be having an acute problem.

Telling them "please wait 2-3 hours until the package is built", while is acceptable, isn't quite ideal / a good look.

I'd love íf there was a way to hint melpa that I just pushed a commit/tag, and that therefore a build is desired.

I reckon that this feature would have some potential for abuse (e.g. someone causing you to repeatedly perform Git ops) so perhaps something like a token or rate limit would be required.

WDYT?

Thanks - V

**Comment by akirak** [comment]
You can clone this repository and build your package locally on your machine. Please follow [an instruction for contributing](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org#preparing-a-pull-request-to-melpa) which can be used to build a package archive installable with `package.el`.

**Comment by vemv** [comment]
Hi, thanks! But I meant something else.

As a package maintainer, sometimes I'd wish for users to perceive bugfixes asap. Maybe it's a particular person, but also, possibly, it's a broader (and unknwn!) audience if the blunder was severe.

Telling an end user to locally install a package seems to be overly tedious for them, and risky.

**Comment by minad** [comment]
FWIW I think MELPA is doing quite well with its ~4 hours waiting time. For comparison, GNU/NonGNU ELPA (stable and devel) recently changed their Cron job to only synchronize the repositories once per day. For me the synchronization happens late in the evening. If I miss pushing just before that time it feels as if I have to wait two days for the package to build, since the package will only be available two mornings later.

**Comment by milkypostman** [comment]
The original design for MELPA had something like this where we had an endpoint on the server that would allow users to trigger a particular build. We then switched to a brute force continuous rebuild system which is more reliable in most ways.

As you point out, we'd be more open to abuse. Also we're trying to minimize toil of managing MELPA and this sort of feature would likely take a bit of work.

With all that said, if you had an idea for a design that would integrate into the current system which was simple and easy to maintain, I can see us at least considering it. If you have thoughts, I think the next step would be filing a fresh bug with a short design summary as it relates to the current system.

I'm sorry I don't have a response that directly solves this problem

**Comment by purcell** [comment]
Yeah, the current scheme is slow but reliable and fair.

In order to reduce the total build cycle time, it's worth noting that 92% of the 5500+ packages are hosted on github. So perhaps there's a way to pre-scan all the github repos using the github API, and then skip unchanged packages faster than performing a no-op `git pull`.

But in general, I doubt that's the bottleneck, because (I believe) we also unconditionally re-build packages even if no updates were pulled. This is kinda useful because it means packages get fixed up automatically if there was temporarily a bug in `package-build`, so we never have to fix stuff up on the server, but it also adds a lot of time. As an example, just building `magit` and its Info documentation takes 29 seconds every build cycle, whether or not the code changed. This is probably where we could work towards improvement, at the cost of some additional complexity.

**Comment by purcell** [comment]
For a small repo (my `exec-path-from-shell`) the cost is more like 7 seconds. We have 8 builders running concurrently, for a total of say 2 hours, 57k build-seconds per build cycle. That's basically an average of 10 seconds per package. We don't have separate timings for fetching vs rebuilding, but we might be able to halve the average time perhaps.

**Comment by purcell** [comment]
At this stage it's probably a no-brainer to also remove the artificial hour-long pause between build cycles, assuming it's fine to have the machine busy 100% of the time, and that we won't start hitting rate limits by fetching all 5000+ repositories even more often!

**Comment by purcell** [comment]
I made the above change, btw, so now there's just a 5 minute pause between the build cycles. Let's see what happens.

**Comment by tarsius** [comment]
*Before I will seriously consider making changes in this direction, I want to make the two new experimental Melpa channels with improved version numbers available. More on that in a bit, in another issue or on a private communication channel.*

----

That being said, I would be happy to comment on some of this.

On the server, we already us `make -k -j8 $(find recipes/ -type f | sort)`, which does wonders. But we don't make that readily available outside of docker. For the longest time (to test changes to `package-build.el`), I used just `make build`, which uses `package-build-archive`, which does *not* build packages in parallel. Doing a full build easily takes more than half a day, using that approach.

I'm now using docker for local testing as well. I've added a few additional targets to `config.mk` to make that simpler, which is particularly useful because I don't use docker elsewhere and thus cannot remember the details. Maybe we should add these targets to the `Makefile` that we ship.

<details>
<summary>My config.mk.</summary>

```make
# Only used by the "pull-package-build" target.
LOAD_PATH = ~/.config/emacs/lib/package-build
# Used in the value of the "EVAL" variable.
PACKAGE_BUILD_REPO ?= "~/.config/emacs/lib/package-build"

docker_build_rebuild:
	docker build -t melpa_builder docker/builder

docker_build_shell:
	docker run -it \
	--mount type=bind,src=$$PWD,target=/mnt/store/melpa \
	-e INHIBIT_MELPA_PULL=t \
	melpa_builder bash

docker_build_run:
	docker run -it \
	--mount type=bind,src=$$PWD,target=/mnt/store/melpa \
	-e INHIBIT_MELPA_PULL=t \
	melpa_builder
```

</details>

----

> In order to reduce the total build cycle time, it's worth noting that 92% of the 5500+ packages are hosted on github. So perhaps there's a way to pre-scan all the github repos using the github API, and then skip unchanged packages faster than performing a no-op `git pull`. I doubt this would be worthwhile, nor particularly easy to do reliably, and the API limits would not be friendly to us when dealing with so many repos.

That's what I do for the Emacsmirror. API limits are not a problem, provided you identify using a token, instead of making anonymous requests. Updating Github packages takes more like 20 minutes (this includes fetching the list and updating packages) that way, instead of two hours. If I ran this more often than about once a week, then it would probably be even faster.

Once it has the list of updated packages, `emir` (sort of Emacsmirror's `package-build`), fetches the packages sequentially. It also does this for packages not hosted on Github, where doing it in parallel does have a real cost, because we cannot first determine which packages actually do have to be updated.

When it comes to pulling from upstream, as the maintainer of both `emir.el` and `package-build.el`, I am more motivated to port the advantages of the latter to the former, than the other way around.

----

There are other advantages that `emir` has over `package-build`, which I have considered porting over.

For example `emir` stores data in an sqlite database. `package-build`'s approach of writing one metadata file per package and then collecting those at the end, seems a bit primitive in comparison.

But I have refrained from such changes for simplicity sake. Generally I have been rather conservative in my changes to `package-build.el`. The code changed a lot, that is true, and there have been some breaking changes, but I have tried to only change *what* it does, when that really was necessary. I concentrated on changing *how* it does thing, with a focus on enabling future changes (such as the mention version string changes), bog fixes, and improving the code quality. The new functionality is mostly opt-in -- at the same time it is limited to specific areas, I would rather avoid having multiple code-paths for various things.

Reasons for why I avoided changes include:

- I didn't have a full picture of what happens on the server side, and was under the impression things were rather static. One reason for that is that I did not want to get involved that aspect of Melpa as well, I already have quite enough on my plate already. At the beginning of the I finally did dive in more, and as a result I now would have some suggestions and also learned some things that could potentially lead to improvements on the elisp side, eventually.

- Melpa is not the only downstream of `package-build`. That includes package authors who submit recipes to Melpa, as we ask them to. But the download numbers indicate, that it is being used for much more than that. Some users may use it directly, others use it as a dependency of other packaging tools. Some are listed at https://github.com/melpa/melpa/wiki/Downstream-Projects, but there might be more. Some of these projects, while popular on their own, where quite stagnant for a long time (but I think this has improved), meaning that any breaking change in `package-build` could affect many users for a long time, until the downstream project made the necessary, usually very minor, change. At least one downstream project even went as far as to bundle an outdated copy of `package-build` in their source.

----

> But in general, I doubt that's the bottleneck, because (I believe) we also unconditionally re-build packages even if no updates were pulled. This is kinda useful because it means packages get fixed up automatically if there was temporarily a bug in `package-build`, so we never have to fix stuff up on the server, ...

That is not necessarily a good thing. Ideally we would rebuild the package version and check if it really did not change. If it did change, then we should bump the version string. Since we don't do that, it can happen, for example, that a file is not included in the package because of a mistake in the recipe, the recipe is fixed and the package is rebuild, and then people get confused why the file is never-the-less from an installation of that very version. (The issue is that the installed the first coming of that version.)

> ... but it also adds a lot of time.

It does, but that's not were I would optimize. If/When we implement support for bumping the version if the recipe and/or `package-build` changed (but not the package that is being build), then we would have to rebuild anyway, so that we can check whether the resulting tarball has changed or not. Or put differently, there is room for a performance gain here, but once we improve other things, we would have to through that out again; so let's not do that.

----

> The original design for MELPA had something like this where we had an endpoint on the server that would allow users to trigger a particular build. We then switched to a brute force continuous rebuild system which is more reliable in most ways.

Interesting, I imagine a minimal viable version would just be a cgi script that runs `make recipe/PROVIDED-PACKAGE-NAME`. The script would come with a risk, but purely on the build side of things, this should just work. We can already run builds in parallel, so triggering another build from elsewhere, doesn't seem like an issue.

Of course that is only the first step. Next we'll be asked to support webhooks, so packages are updates as soon as new commits become available. :grin: 

(See https://docs.github.com/en/webhooks/about-webhooks.)

**Comment by purcell** [comment]
Yeah, thanks for sharing those thoughts. Simple is good. The github API thing would be interesting though. Webhooks and CGI scripts, less so. It's kinda great that everything is static currently.

The docker config in here is a bit special-case — it assumes certain mount points, and you can't just "docker compose up" to start your own local MELPA, which would be ideal in order to ease contributions and testing.

**Comment by tarsius** [comment]
> The docker config in here is a bit special-case — it assumes certain mount points, and you can't just "docker compose up" to start your own local MELPA, which would be ideal in order to ease contributions and testing.

I actually found it fairly easy to setup (for the build image, didn't try the others). `--mount type=bind,src=$$PWD,target=/mnt/store/melpa` is really all there is to it.

Well, there is at least one more complication: Inside docker root is used, and because I am using the same checkout outside of docker, I have to fixup permissions afterwards. Could we change this to "root-less docker"?

**Comment by purcell** [comment]
Potentially. It's all root on the server. I'm not sure what the steps would be to switch over.

**Comment by vemv** [comment]
> I made the above change, btw, so now there's just a 5 minute pause between the build cycles. Let's see what happens.

I think I have since seen an improvement in release times. If that wasn't placebo effect, thanks for that!
